### PR TITLE
flash 2.7.0 (new formula)

### PR DESCRIPTION
**Pull request by witsch**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Review by chenrui333 — APPROVED** [review]

**Comment by witsch** [comment]
Hmm, the "build bottles" step [failed for 10.15](https://github.com/Homebrew/homebrew-core/pull/52503/checks?check_run_id=558569838) because of "GitHub API Error: API rate limit exceeded". I guess that should be okay, though, because this [doesn't need a bottle anyway](https://github.com/Homebrew/homebrew-core/pull/52503/files#diff-300878c5e53110263517c49649533be7R7).  Or it there any way to trigger a re-run?

File: Formula/flash.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+  bottle :unneeded
+
+  def install
+    bin.install "flash"
+  end
+
+  test do
+    # Unfortunately there's not much else to test without actually
+    # downloading an image and/or writing it to an SD card
```
Comment (SMillerDev) [inline Formula/flash.rb:15]: You can download an image using a resource block.

Comment (witsch) [reply Formula/flash.rb:15]: I can try to add a test like that, but the script then waits for the user to insert an SD card.  That is, unless I could mock `diskutil` somehow [like in flash's own Bats tests](https://github.com/hypriot/flash/blob/2.6.0/test/test_helper.bash#L19)?

Comment (witsch) [reply Formula/flash.rb]: Ok, d4eb3b4aa36b8e93c92568a0ade97057767ad07b has a "real" test.  It mocks a few of the used binaries, but now it actually does something.  I also tried to add some helpful comments, but I'm not sure it's too specific regarding the actual implementation…

File: Formula/flash.rb [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+  bottle :unneeded
+
+  def install
+    bin.install "flash"
+  end
+
+  test do
+    # mock required binaries...
+    ENV.prepend_path "PATH", testpath
```
Comment (SMillerDev) [inline Formula/flash.rb:15]: Mocking all functional parts is not a great idea. It might be better to test for failure then.

Comment (witsch) [reply Formula/flash.rb]: Ok, how about ae4c5b18675e122d6f1bb60e0922ad2ff75da738 then? 🙂

File: Formula/flash.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    bin.install "flash"
+  end
+
+  test do
+    system "hdiutil", "create", "-size", "128k", "test.dmg"
+    command = "#{bin}/flash --force --device /dev/disk42 test.dmg"
+    stdout, stderr, status = Open3.capture3(command)
+    assert_equal false, status.success?
+    assert_equal "", stdout
+    assert_match "Could not find disk: /dev/disk42", stderr
```
Comment (SMillerDev) [inline Formula/flash.rb:19]: This is the homebrew way.
```suggestion
    output = shell_output("#{bin}/flash --force --device /dev/disk42 test.dmg 2>&1")
    assert_match "Could not find disk: /dev/disk42", output
```

Comment (witsch) [reply Formula/flash.rb]: Hi Sean,

> This is the homebrew way.

But somehow it doesn't seem to work reliably… 😞

Incidentally, I've actually tried with `shell_output` and various combination of '1', '2', '&' & '>' before resorting to `capture3`, but thought I had mixed up the order again and gave up.  Now it's working locally for me, but apparently not on all versions of OSX.  How do we proceed?

**Comment by SMillerDev** [comment]
Fails on everything below catalina

**Comment by witsch** [comment]
> Fails on everything below catalina

Yes, it does, but I don't know how to fix it (also see [my comment above](https://github.com/Homebrew/homebrew-core/pull/52503#discussion_r405036507)).  It worked using `capture3` (except for the github rate limit), so should we perhaps just revert the last change?

**Comment by witsch** [comment]
> > Fails on everything below catalina
> 
> Yes, it does, but I don't know how to fix it (also see [my comment above](https://github.com/Homebrew/homebrew-core/pull/52503#discussion_r405036507)). It worked using `capture3` (except for the github rate limit), so should we perhaps just revert the last change?

Ok, or perhaps it didn't… 😞  I'll look into it a bit later then.

**Comment by SMillerDev** [comment]
Using open3 is not the solution, there are a lot of tests using the homebrew aliasses that work fine.

**Comment by witsch** [comment]
> Using open3 is not the solution, there are a lot of tests using the homebrew aliasses that work fine.

Yes, you're right.  I've just found out that the problem is really `diskutil` behaving differently starting from Catalina.  There it outputs `Could not find disk: /dev/disk42`, at least on Mojave is fails silently… 🙄  I'll try to find something that works on both.

**Comment by witsch** [comment]
> […] I'll try to find something that works on both.

I've changed the test to answer the [`read` prompt](https://github.com/hypriot/flash/blob/2.7.0/flash#L627) with a bogus answer. This avoids the issue with `diskutil` outputting errors differently.  The prompt itself cannot be tested, because it's only output in tty mode...

**Comment by witsch** [comment]
> Using open3 is not the solution, there are a lot of tests using the homebrew aliasses that work fine.

Hi Sean, would you mind having another look?  I think the test should be okay now and it's finally passing on all OSX versions… 🙂

**Comment by SMillerDev** [comment]
Looks good, please squash all commits into one `<name> <version> (new formula)` commit.

**Comment by witsch** [comment]
> Looks good, please squash all commits into one `<name> <version> (new formula)` commit.

Done 🙂  And thanks for helping with the test!

**Comment by SMillerDev** [comment]
Thanks @witsch! Without contributions like yours it'd be impossible to keep homebrew going with the high standards that users have come to expect from the project. You can feel good knowing that you've made the world a tiny bit better for homebrew users around the world! 👍 🎉

**Pull request closed without merging**